src/contract.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from enums import *
+import numpy as np
+
+
+class Contract(ABC):
+    _underlying: Stock
+    _derivative_type: PutCallFwd
+    _strike: float
+    _expiry: float
+
+    def get_und(self) -> Stock:
+        return self._underlying
+
+    def get_type(self) -> PutCallFwd:
+        return self._derivative_type
+
+    def get_strike(self) -> float:
+        return self._strike
+
+    def get_expiry(self) -> float:
+        return self._expiry
+
+    @abstractmethod
+    def display(self) -> dict[str, any]:
+        pass
+
+    @abstractmethod
+    def convert_to_generic(self) -> GenericContract:
+        pass
+
+    @abstractmethod
+    def get_timeline(self) -> list[float]:
+        pass
+
+
+class VanillaContract(Contract):
+    @abstractmethod
+    def convert_to_generic(self) -> GenericContract:
+        pass
+
+    @abstractmethod
+    def payoff(self, spot: float) -> float:
+        pass
+
+    def __init__(self, und: Stock, dtype: PutCallFwd, strk: float, exp: float) -> None:
+        self._underlying = und
+        self._derivative_type = dtype
+        self._strike = strk
+        self._expiry = exp
+
+    def display(self) -> dict[str, any]:
```

Comment:
I would put this into the __str__ method. So any time the contract is casted as string, it will report your desired form.

Reply:
To be honest, I also wanted to use __str__ for this, but you can't force implementation of predefined methods as they already have a default implementation.

Reply:
I would try something like adding __str__ with a NotImplemented exception inside to the parent class. It worked for me before with simple class methods, maybe worth a try.